CombinatorialGames/Game/Concrete.lean

```diff
@@ -0,0 +1,146 @@
+
+/-- Use `toLeftMovesPGame` to cast between these two types. -/
+theorem leftMoves_toPGame (a : α) : (toPGame a).LeftMoves = {b // b ≺ₗ a} := by
+  rw [toPGame]; rfl
+/-- Use `toRightMovesPGame` to cast between these two types. -/
+theorem rightMoves_toPGame (a : α) : (toPGame a).RightMoves = {b // b ≺ᵣ a} := by
+  rw [toPGame]; rfl
+
+theorem moveLeft_toPGame_hEq (a : α) :
+    HEq (toPGame a).moveLeft fun i : {b // b ≺ₗ a} => toPGame i.1 := by
+  rw [toPGame]; rfl
+theorem moveRight_toPGame_hEq (a : α) :
+    HEq (toPGame a).moveRight fun i : {b // b ≺ᵣ a} => toPGame i.1 := by
+  rw [toPGame]; rfl
+
+/-- Turns a left-subsequent position for `a` into a left move for `toPGame a` and vice versa.
+
+Even though these types are the same (not definitionally so), this is the preferred way to convert
+between them. -/
+def toLeftMovesPGame {a : α} : {b // b ≺ₗ a} ≃ (toPGame a).LeftMoves :=
+  Equiv.cast (leftMoves_toPGame a).symm
+/-- Turns a right-subsequent position for `a` into a right move for `toPGame a` and vice versa.
+
+Even though these types are the same (not definitionally so), this is the preferred way to convert
+between them. -/
+def toRightMovesPGame {a : α} : {b // b ≺ᵣ a} ≃ (toPGame a).RightMoves :=
+  Equiv.cast (rightMoves_toPGame a).symm
+
+@[simp]
+theorem moveLeft_toPGame {a : α} (i) :
+    (toPGame a).moveLeft i = toPGame (toLeftMovesPGame.symm i).1 :=
+  (congr_heq (moveLeft_toPGame_hEq a).symm (cast_heq _ i)).symm
+@[simp]
+theorem moveRight_toPGame {a : α} (i) :
+    (toPGame a).moveRight i = toPGame (toRightMovesPGame.symm i).1 :=
+  (congr_heq (moveRight_toPGame_hEq a).symm (cast_heq _ i)).symm
+
+theorem moveLeft_toPGame_toLeftMovesPGame {a : α} (i) :
+    (toPGame a).moveLeft (toLeftMovesPGame i) = toPGame i.1 :=
+  by simp
+theorem moveRight_toPGame_toRightMovesPGame {a : α} (i) :
+    (toPGame a).moveRight (toRightMovesPGame i) = toPGame i.1 :=
+  by simp
+
+@[simp]
+theorem toLeftMovesPGame_symm_prop {a : α} (i : (toPGame a).LeftMoves) :
+    (toLeftMovesPGame.symm i).1 ≺ₗ a :=
+  (toLeftMovesPGame.symm i).prop
+@[simp]
+theorem toRightMovesPGame_symm_prop {a : α} (i : (toPGame a).RightMoves) :
+    (toRightMovesPGame.symm i).1 ≺ᵣ a :=
+  (toRightMovesPGame.symm i).prop
+
+-- TODO: PR to Mathlib
+theorem heq_subtype {α : Type*} {p q : α → Prop} {a : Subtype p} {b : Subtype q} (h : p = q)
+    (he : HEq a b) : a.1 = b.1 := by
+  subst h
+  simpa [Subtype.eq_iff] using he
+
+theorem neg_toPGame (h : subsequentL (α := α) = subsequentR) (a : α) : -toPGame a = toPGame a := by
+  rw [toPGame, neg_def]
+  congr
+  all_goals try (ext; rw [h])
+  all_goals
+    apply Function.hfunext (by rw [h])
+    simp_rw [heq_eq_eq, Subtype.forall, h]
+    intro a ha b hb he
+    have : a = b := heq_subtype (by rw [h]) he
+    subst this
+    have := subrelation_subsequentR hb
+    apply neg_toPGame h
+termination_by isWellFounded_subsequent.wf.wrap a
+
+theorem impartial_toPGame (h : subsequentL (α := α) = subsequentR) (a : α) :
```

Comment:
Should this theorem be here, or in the impartial file?

Reply:
The theorem being here seems fine: I can see many impartial games being defined, and this theorem is small enough that it doesn't warrant some ConcreteImpartial file.

Reply:
It can stay here: Impartial games are common enough that this theorem can stay in Concrete for now, though I could see a future where `ImpartialConcrete` is realized.